### PR TITLE
Add configurable options to the AppEngine object.

### PR DESCRIPTION
**Pull request by rrch**
Use an http.Agent for all api and metadata calls.
Default maxSockets for the agent to 100 (https://github.com/GoogleCloudPlatform/appengine-nodejs/issues/34).

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -42,11 +42,19 @@ var http = require('http');
...
 var winston = require('winston');
 
-function AppEngine() {
+function AppEngine(options) {
+  this.options_ = options || {};
   this.serializer_ = new net.proto2.contrib.WireSerializer();
   this.cache_ = {};
-  this.logSizeThreshold_ = 1024 * 1024; // 1 MB
-  this.logMaxTimestampDeltaMillis_ = 60000; // 60 seconds
+  this.defaults_ = {
```
Comment (rakyll) [inline lib/index.js]: Would be lovely if we remove the defaults from the prototype.

File: lib/index.js [hunk 2]
```diff
@@ -42,11 +42,19 @@ var http = require('http');
...
+    logSizeThreshold: 1024 * 1024, // 1 MB
+    logMaxTimestampDeltaMillis: 60000, // 60 seconds
+    loggerMaxFiles: 1,
+    loggerMaxFileSize: 100 * 1024 * 1024, // 100 MB
+    httpAgentMaxSockets: 100
+  };
+  this.logSizeThreshold_ = this.options_.logSizeThreshold || this.defaults_.logSizeThreshold;
```
Comment (rakyll) [inline lib/index.js]: We need a function that  overrides the values of the defaults with the values of the options.

Requiring user to rely on `options_.v || defaults_.v` is not great for readability in the longer term.

File: lib/index.js (hunk 1 above)
Comment (rrch) [inline lib/index.js]: Did you mean "remove from" or "move to"? E.g.
AppEngine.prototype.defaultOptions_ = { foo: 1, bar: 2 };

**Comment by rrch** [comment]
PTAL

File: lib/index.js [hunk 3]
```diff
@@ -58,13 +60,28 @@ function AppEngine() {
...
+  }
 }
 
+AppEngine.prototype.defaultOptions_ = {
+  logSizeThreshold: 1024 * 1024, // 1 MB
+  logMaxTimestampDeltaMillis: 60000, // 60 seconds
+  loggerMaxFiles: 1,
+  loggerMaxFileSize: 100 * 1024 * 1024, // 100 MB
+  httpAgentMaxSockets: 100,
+  exportPublicFunctions: true
```
Comment (rakyll) [inline lib/index.js]: We should add a comment it's building the surface layer. I'd go with different name such as initialize or build or whatever as we discussed earlier.

File: lib/index.js [hunk 4]
```diff
@@ -986,9 +1006,7 @@ function makeModulesGetHostnameProto(module, version, instance) {
 }
 
 function createAppEngine() {
-  var ae = new AppEngine();
-  ae.exportAll_();
-  return ae;
+  return new AppEngine();
 }
 
 module.exports = createAppEngine();
```
Comment (rakyll) [inline lib/index.js]: module.exports = new AppEngine();

File: lib/index.js [hunk 5]
```diff
@@ -986,9 +1006,7 @@ function makeModulesGetHostnameProto(module, version, instance) {
 }
 
 function createAppEngine() {
```
Comment (rakyll) [inline lib/index.js]: Remove.

File: test/index.js [hunk 6]
```diff
@@ -49,9 +49,84 @@ describe('appengine', function() {
...
+      });
+
+      function verifyAllOptionsHaveDefaultValues(ae) {
+        assert.strictEqual(ae.logSizeThreshold_, ae.defaultOptions_.logSizeThreshold);
+        assert.strictEqual(ae.logMaxTimestampDeltaMillis_, ae.defaultOptions_.logMaxTimestampDeltaMillis);
+        assert.strictEqual(ae.httpAgent_.maxSockets, ae.defaultOptions_.httpAgentMaxSockets);
+        var transport = ae.logger_.transports.file;
+        assert.ok(!!transport);
+        assert.strictEqual(transport.maxFiles, ae.defaultOptions_.loggerMaxFiles);
+        assert.strictEqual(transport.maxsize, ae.defaultOptions_.loggerMaxFileSize);
```
Comment (rakyll) [inline test/index.js]: Add an assertion for the exportPublicFunctions.

File: test/index.js [hunk 7]
```diff
@@ -475,6 +550,21 @@ describe('appengine', function() {
         });
       });
 
+      it('uses getRemoteApiRequestOptions_ to get the request options and passes them to sendHttpRequest_', function(done) {
+        var ae = new appengine.AppEngine();
+        var requestOptions = {};
```
Comment (rakyll) [inline test/index.js]: Shouldn't we primarily test it with a non-empty object?

File: lib/index.js (hunk 3 above)
Comment (rrch) [inline lib/index.js]: Done. I've renamed it to "addPublicAPIs" and I've also renamed exportAll_ to addPublicAPIs_.

File: lib/index.js (hunk 4 above)
Comment (rrch) [inline lib/index.js]: Done

File: lib/index.js (hunk 5 above)
Comment (rrch) [inline lib/index.js]: Done

File: test/index.js (hunk 6 above)
Comment (rrch) [inline test/index.js]: Done

File: test/index.js (hunk 7 above)
Comment (rrch) [inline test/index.js]: We're using strictEqual in the test, so only the identity of the object matters.

**Comment by rakyll** [comment]
LGTM

**Pull request merged**